<img src="http://imgur.com/1ZcRyrc.png" style="float: left; margin: 20px; height: 55px">

## Practice Loading and Describing Data 

_Authors: Matt Brems (DC)_

---

In this lab you will practice loading data using python and describing it with statistics.

It might be a good idea to first check the [source of the Boston housing data](https://archive.ics.uci.edu/ml/datasets/Housing).

### 1. Load the boston housing data (provided)

In [75]:
# Download the data and save to a file called "housing.data."
from pprint import pprint
import urllib.request
data_url = "https://archive.ics.uci.edu/ml/machine-learning-databases/housing/housing.data"

# this saves a file called 'housing.data' locally'
urllib.request.urlretrieve(data_url, './datasets/housing.data')

('./datasets/housing.data', <http.client.HTTPMessage at 0x117525e10>)

The data file does not contain the column names in the first line, so we'll need to add those in manually. You can find the names and explanations [here](https://archive.ics.uci.edu/ml/machine-learning-databases/housing/housing.names). We've extracted the names below for your convenience. You may choose to edit the names, should you decide it would be more helpful to do so.

In [76]:
names = ["CRIM", "ZN", "INDUS", "CHAS", "NOX", "RM", "AGE",
         "DIS", "RAD", "TAX", "PTRATIO", "B", "LSTAT", "MEDV"]

### 2. Load the `housing.data` file with python

Using any method of your choice.

> _**Hint:** despite this file having a strange `.data` extension, using python's `open() as file` and `file.read()` or `file.readlines()` we can load this in and see that it is a text file formatted much the same as a CSV. You can use string operations to format the data._

In [89]:
housing_data = '/Users/ajinthchristudas/GA/Labs/lab-statistics-describing_data/datasets/housing.data'
housing_data_list = []

with open(housing_data, 'r') as f:
    for line in f: 
        split_line = (line).split()
        housing_data_list.append(split_line)

pprint (housing_data_list[0:1])

[['0.00632',
  '18.00',
  '2.310',
  '0',
  '0.5380',
  '6.5750',
  '65.20',
  '4.0900',
  '1',
  '296.0',
  '15.30',
  '396.90',
  '4.98',
  '24.00']]


In [91]:
'''Converting the list to Pandas Data Frame '''
housing_df = pd.DataFrame.from_records(housing_data_list, columns=names)
housing_df.head()

CRIM     ZN  INDUS CHAS     NOX      RM    AGE     DIS RAD    TAX  \
0  0.00632  18.00  2.310    0  0.5380  6.5750  65.20  4.0900   1  296.0   
1  0.02731   0.00  7.070    0  0.4690  6.4210  78.90  4.9671   2  242.0   
2  0.02729   0.00  7.070    0  0.4690  7.1850  61.10  4.9671   2  242.0   
3  0.03237   0.00  2.180    0  0.4580  6.9980  45.80  6.0622   3  222.0   
4  0.06905   0.00  2.180    0  0.4580  7.1470  54.20  6.0622   3  222.0   

  PTRATIO       B LSTAT   MEDV  
0   15.30  396.90  4.98  24.00  
1   17.80  396.90  9.14  21.60  
2   17.80  392.83  4.03  34.70  
3   18.70  394.63  2.94  33.40  
4   18.70  396.90  5.33  36.20

### 3.  Conduct a brief integrity check of your data. 

This integrity check should include, but is not limited to, checking for missing values and making sure all values make logical sense. (i.e. is one variable a percentage, but there are observations above 100%?)

Summarize your findings in a few sentences, including what you checked and, if appropriate, any 
steps you took to rectify potential integrity issues.

In [94]:
housing_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 506 entries, 0 to 505
Data columns (total 14 columns):
CRIM       506 non-null object
ZN         506 non-null object
INDUS      506 non-null object
CHAS       506 non-null object
NOX        506 non-null object
RM         506 non-null object
AGE        506 non-null object
DIS        506 non-null object
RAD        506 non-null object
TAX        506 non-null object
PTRATIO    506 non-null object
B          506 non-null object
LSTAT      506 non-null object
MEDV       506 non-null object
dtypes: object(14)
memory usage: 55.4+ KB


In [95]:
housing.dtypes

CRIM       float64
ZN         float64
INDUS      float64
CHAS       float64
NOX        float64
RM         float64
AGE        float64
DIS        float64
RAD        float64
TAX        float64
PTRATIO    float64
B          float64
LSTAT      float64
MEDV       float64
dtype: object

### 4. For what two attributes does it make the *least* sense to calculate mean and median? Why?

In [97]:
# A: 
housing_df.head()

CRIM     ZN  INDUS CHAS     NOX      RM    AGE     DIS RAD    TAX  \
0  0.00632  18.00  2.310    0  0.5380  6.5750  65.20  4.0900   1  296.0   
1  0.02731   0.00  7.070    0  0.4690  6.4210  78.90  4.9671   2  242.0   
2  0.02729   0.00  7.070    0  0.4690  7.1850  61.10  4.9671   2  242.0   
3  0.03237   0.00  2.180    0  0.4580  6.9980  45.80  6.0622   3  222.0   
4  0.06905   0.00  2.180    0  0.4580  7.1470  54.20  6.0622   3  222.0   

  PTRATIO       B LSTAT   MEDV  
0   15.30  396.90  4.98  24.00  
1   17.80  396.90  9.14  21.60  
2   17.80  392.83  4.03  34.70  
3   18.70  394.63  2.94  33.40  
4   18.70  396.90  5.33  36.20

* CHAS (Charles River Dummy Variable) column is binary variable hence finding the mean and median for a column that houses the binary variable does not make a whole lot of sense

### 5. Which two variables have the strongest linear association? 

Report both variables, the metric you chose as the basis for your comparison, and the value of that metric. *(Hint: Make sure you consider only variables for which it makes sense to find a linear association.)*

In [6]:
# A:

### 6. Look at distributional qualities of variables.

Answer the following questions:
1. Which variable has the most symmetric distribution? 
2. Which variable has the most left-skewed (negatively skewed) distribution? 
3. Which variable has the most right-skewed (positively skewed) distribution? 

Defend your method for determining this.

In [7]:
# A:

### 8. Repeat question 6 but scale the variables by their range first.

As you may have noticed, the spread of the distribution contributed significantly to the results in question 6.

In [8]:
# A:

### 9. Univariate analysis of your choice

Conduct a full univariate analysis on MEDV, CHAS, TAX, and RAD. 

For each variable, you should answer the three questions generally asked in a univariate analysis using the most appropriate metrics.
- A measure of central tendency
- A measure of spread
- A description of the shape of the distribution (plot or metric based)

If you feel there is additional information that is relevant, include it. 

In [9]:
# A:

### 10. Reducing the number of observations

It seems likely that this data is a census - that is, the data set includes the entire target population. Suppose that the 506 observations was too much for our computer (as unlikely as this might be) and we needed to pare this down to fewer observations. 

**10.A Use the `random.sample()` function to select 50 observations from `'AGE'`.**

([This documentation](https://docs.python.org/2/library/random.html) may be helpful.)

In [11]:
# A:

**10.B Identify the type of sampling we just used.**

In [12]:
# A: